File: src-tauri/src/core/import/swagger.rs
use crate::core::types::{
    AuthConfig, AuthType, Endpoint, EndpointMetadata, EnvironmentConfig, NameValue,
    PreflightConfig, Service,
};
use openapiv3::OpenAPI;
use std::time::{SystemTime, UNIX_EPOCH};

pub fn parse_spec_content(
    content: &str,
    service_id: &str,
) -> Result<(String, Vec<Endpoint>), String> {
    let value: serde_json::Value = if content.trim().starts_with('{') {
        serde_json::from_str(content).map_err(|e| format!("Failed to parse JSON: {}", e))?
    } else {
        serde_yaml::from_str(content).map_err(|e| format!("Failed to parse YAML: {}", e))?
    };

    let mut endpoints = Vec::new();
    let base_url;

    if value.get("openapi").is_some() {
        let openapi: OpenAPI =
            serde_json::from_value(value).map_err(|e| format!("Invalid OpenAPI 3: {}", e))?;

        base_url = openapi
            .servers
            .first()
            .map(|s| s.url.clone())
            .unwrap_or_else(|| "https://api.example.com".to_string());

        for (path, path_item) in openapi.paths.iter() {
            if let Some(item) = path_item.as_item() {
                let methods = [
                    ("GET", &item.get),
                    ("POST", &item.post),
                    ("PUT", &item.put),
                    ("DELETE", &item.delete),
                    ("PATCH", &item.patch),
                ];

                for (method, op_opt) in methods {
                    if let Some(op) = op_opt {
                        let endpoint_id = format!("e-{}", uuid::Uuid::new_v4());
                        let endpoint_name = op
                            .summary
                            .clone()
                            .or_else(|| op.operation_id.clone())
                            .unwrap_or_else(|| format!("{} {}", method, path));

                        let mut params = Vec::new();
                        let mut headers = Vec::new();

                        for param_ref in &op.parameters {
                            if let Some(p) = param_ref.as_item() {
                                match p {
                                    openapiv3::Parameter::Query { parameter_data, .. } => {
                                        params.push(NameValue {
                                            name: parameter_data.name.clone(),
                                            value: "".to_string(),
                                            enabled: true,
                                            secret_key: None,
                                        });
                                    }
                                    openapiv3::Parameter::Header { parameter_data, .. } => {
                                        headers.push(NameValue {
                                            name: parameter_data.name.clone(),
                                            value: "".to_string(),
                                            enabled: true,
                                            secret_key: None,
                                        });
                                    }
                                    _ => continue,
                                }
                            }
                        }

                        endpoints.push(create_endpoint(
                            endpoint_id,
                            service_id.to_string(),
                            endpoint_name,
                            method.to_string(),
                            path.clone(),
                            params,
                            headers,
                        ));
                    }
                }
            }
        }
    } else {
        let host = value
            .get("host")
            .and_then(|v| v.as_str())
            .unwrap_or("api.example.com");
        let base_path = value.get("basePath").and_then(|v| v.as_str()).unwrap_or("");
        let scheme = value
            .get("schemes")
            .and_then(|v| v.as_array())
            .and_then(|a| a.first())
            .and_then(|v| v.as_str())
            .unwrap_or("https");

        base_url = format!("{}://{}{}", scheme, host, base_path);

        if let Some(paths) = value.get("paths").and_then(|v| v.as_object()) {
            for (path, path_value) in paths {
                if let Some(methods_obj) = path_value.as_object() {
                    for (method, op_value) in methods_obj {
                        let method_upper = method.to_uppercase();
                        if !["GET", "POST", "PUT", "DELETE", "PATCH"]
                            .contains(&method_upper.as_str())
                        {
                            continue;
                        }

                        let endpoint_id = format!("e-{}", uuid::Uuid::new_v4());
                        let endpoint_name = op_value
                            .get("summary")
                            .and_then(|v| v.as_str())
                            .or_else(|| op_value.get("operationId").and_then(|v| v.as_str()))
                            .map(|s| s.to_string())
                            .unwrap_or_else(|| format!("{} {}", method_upper, path));

                        let mut params = Vec::new();
                        let mut headers = Vec::new();

                        if let Some(parameters) =
                            op_value.get("parameters").and_then(|v| v.as_array())
                        {
                            for p in parameters {
                                let p_name = p.get("name").and_then(|v| v.as_str()).unwrap_or("");
                                let p_in = p.get("in").and_then(|v| v.as_str()).unwrap_or("");

                                if p_in == "query" {
                                    params.push(NameValue {
                                        name: p_name.to_string(),
                                        value: "".to_string(),
                                        enabled: true,
                                        secret_key: None,
                                    });
                                } else if p_in == "header" {
                                    headers.push(NameValue {
                                        name: p_name.to_string(),
                                        value: "".to_string(),
                                        enabled: true,
                                        secret_key: None,
                                    });
                                }
                            }
                        }

                        endpoints.push(create_endpoint(
                            endpoint_id,
                            service_id.to_string(),
                            endpoint_name,
                            method_upper,
                            path.clone(),
                            params,
                            headers,
                        ));
                    }
                }
            }
        }
    }

    Ok((base_url, endpoints))
}

/// Create a new Service from a parsed spec with default environments
pub fn create_service_from_spec(
    name: String,
    directory: String,
    base_url: String,
    endpoints: Vec<Endpoint>,
) -> Service {
    let service_id = format!(
        "s-{}",
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis()
    );

    Service {
        id: service_id,
        name,
        environments: vec![
            EnvironmentConfig {
                name: "DEV".to_string(),
                is_unsafe: false,
                variables: vec![NameValue {
                    name: "BASE_URL".to_string(),
                    value: base_url.clone(),
                    enabled: true,
                    secret_key: None,
                }],
            },
            EnvironmentConfig {
                name: "STAGE".to_string(),
                is_unsafe: false,
                variables: vec![NameValue {
                    name: "BASE_URL".to_string(),
                    value: base_url.clone(),
                    enabled: true,
                    secret_key: None,
                }],
            },
            EnvironmentConfig {
                name: "PROD".to_string(),
                is_unsafe: true,
                variables: vec![NameValue {
                    name: "BASE_URL".to_string(),
                    value: base_url,
                    enabled: true,
                    secret_key: None,
                }],
            },
        ],
        is_authenticated: false,
        auth_type: Some(AuthType::None),
        auth: AuthConfig {
            r#type: "none".to_string(),
            active: true,
            basic_user: "".to_string(),
            basic_pass: "".to_string(),
            bearer_token: "".to_string(),
            api_key_name: "".to_string(),
            api_key_value: "".to_string(),
            api_key_location: "header".to_string(),
        },
        preflight: PreflightConfig {
            enabled: false,
            method: "POST".to_string(),
            url: "".to_string(),
            body: "".to_string(),
            body_type: "application/json".to_string(),
            body_params: vec![],
            headers: vec![],
            cache_token: true,
            cache_duration: "".to_string(),
            cache_duration_key: "".to_string(),
            cache_duration_unit: "seconds".to_string(),
            token_key: "".to_string(),
            token_header: None,
        },
        endpoints,
        directory,
        selected_environment: Some("DEV".to_string()),
        git_url: None,
    }
}

fn create_endpoint(
    id: String,
    service_id: String,
    name: String,
    method: String,
    url: String,
    params: Vec<NameValue>,
    headers: Vec<NameValue>,
) -> Endpoint {
    Endpoint {
        id,
        service_id,
        name,
        method,
        url,
        authenticated: false,
        auth_type: "none".to_string(),
        metadata: EndpointMetadata {
            version: "1.0".to_string(),
            last_updated: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        },
        params,
        headers,
        body: "".to_string(),
        preflight: PreflightConfig {
            enabled: false,
            method: "GET".to_string(),
            url: "".to_string(),
            body: "".to_string(),
            body_type: "application/json".to_string(),
            body_params: vec![],
            headers: vec![],
            cache_token: true,
            cache_duration: "derived".to_string(),
            cache_duration_key: "expires_in".to_string(),
            cache_duration_unit: "seconds".to_string(),
            token_key: "access_token".to_string(),
            token_header: Some("Authorization".to_string()),
        },
        last_version: 0,
        versions: vec![],
    }
}

File: internal/import/swagger.go
package importlib

import (
	"fmt"
	"strings"
	"time"

	"xrest/internal/models"

	"github.com/getkin/kin-openapi/openapi2"
	"github.com/getkin/kin-openapi/openapi2conv"
	"github.com/getkin/kin-openapi/openapi3"
	"github.com/google/uuid"
	"gopkg.in/yaml.v3"
)

// ParseSpecContent parses an OpenAPI 3 or Swagger 2 spec and returns base URL + endpoints.
func ParseSpecContent(content, serviceID string) (string, []models.Endpoint, error) {
	if strings.TrimSpace(content) == "" {
		return "", nil, fmt.Errorf("empty spec content")
	}

	now := uint64(time.Now().Unix())

	// Try as OpenAPI 3 JSON
	{
		doc3 := &openapi3.T{}
		if err := doc3.UnmarshalJSON([]byte(content)); err == nil {
			return doc3ToEndpoints(doc3, serviceID, now)
		}
	}

	// Try as OpenAPI 3 YAML
	{
		loader := &openapi3.Loader{}
		doc3, err := loader.LoadFromData([]byte(content))
		if err == nil && doc3 != nil && doc3.Info != nil {
			return doc3ToEndpoints(doc3, serviceID, now)
		}
	}

	// Try as Swagger 2 JSON
	{
		doc2 := &openapi2.T{}
		if err := doc2.UnmarshalJSON([]byte(content)); err == nil && doc2.Swagger != "" {
			doc3, err := openapi2conv.ToV3(doc2)
			if err == nil {
				return doc3ToEndpoints(doc3, serviceID, now)
			}
		}
	}

	// Try as Swagger 2 YAML
	{
		doc2 := &openapi2.T{}
		if err := yaml.Unmarshal([]byte(content), doc2); err == nil && doc2.Swagger != "" {
			doc3, err := openapi2conv.ToV3(doc2)
			if err == nil {
				return doc3ToEndpoints(doc3, serviceID, now)
			}
		}
	}

	// Fallback: parse as generic map and handle manually
	var spec map[string]interface{}
	if err := yaml.Unmarshal([]byte(content), &spec); err != nil {
		return "", nil, fmt.Errorf("failed to parse spec as YAML: %w", err)
	}
	return parseGenericSpec(spec, serviceID, now)
}

// doc3ToEndpoints extracts endpoints from a parsed OpenAPI 3 document.
func doc3ToEndpoints(doc3 *openapi3.T, serviceID string, now uint64) (string, []models.Endpoint, error) {
	var baseURL string
	if len(doc3.Servers) > 0 {
		baseURL = doc3.Servers[0].URL
	} else {
		baseURL = "https://api.example.com"
	}

	var endpoints []models.Endpoint

	if doc3.Paths == nil {
		return baseURL, endpoints, nil
	}

	for _, path := range doc3.Paths.InMatchingOrder() {
		pi := doc3.Paths.Value(path)
		if pi == nil {
			continue
		}

		ops := map[string]*openapi3.Operation{
			"GET":     pi.Get,
			"POST":    pi.Post,
			"PUT":     pi.Put,
			"DELETE":  pi.Delete,
			"PATCH":   pi.Patch,
			"HEAD":    pi.Head,
			"OPTIONS": pi.Options,
		}

		for method, op := range ops {
			if op == nil {
				continue
			}
			endpointName := op.Summary
			if endpointName == "" {
				endpointName = op.OperationID
			}
			if endpointName == "" {
				endpointName = fmt.Sprintf("%s %s", method, path)
			}

			var params []models.Param
			var headers []models.Header

			for _, paramRef := range op.Parameters {
				if paramRef == nil || paramRef.Value == nil {
					continue
				}
				p := paramRef.Value
				nv := models.NameValue{
					Name:    p.Name,
					Value:   "",
					Enabled: true,
					Type:    "plain",
				}
				if p.Schema != nil && p.Schema.Value != nil {
					if defaultVal, ok := p.Schema.Value.Default.(string); ok {
						nv.Value = defaultVal
					}
				}
				switch p.In {
				case "query":
					params = append(params, nv)
				case "header":
					headers = append(headers, nv)
				}
			}

			endpoints = append(endpoints, models.Endpoint{
				ID:        "e-" + uuid.NewString(),
				ServiceID: serviceID,
				Name:      endpointName,
				Method:    method,
				URL:       path,
				Metadata: models.EndpointMetadata{
					Version:     "1.0",
					LastUpdated: now,
				},
				Params:      params,
				Headers:     headers,
				Preflight:   nil,
				LastVersion: 0,
				Versions:    nil,
			})
		}
	}

	return baseURL, endpoints, nil
}

// parseGenericSpec handles manually-parsed specs.
func parseGenericSpec(spec map[string]interface{}, serviceID string, now uint64) (string, []models.Endpoint, error) {
	baseURL := extractBaseURL(spec)

	var endpoints []models.Endpoint

	paths, ok := getMap(spec, "paths")
	if !ok {
		return "", nil, fmt.Errorf("spec has no 'paths' field")
	}

	for path, pathItem := range paths {
		pathStr, ok := pathItem.(map[string]interface{})
		if !ok {
			continue
		}
		for method, opValue := range pathStr {
			methodUpper := strings.ToUpper(method)
			if !isHTTPMethod(methodUpper) {
				continue
			}
			opMap, ok := opValue.(map[string]interface{})
			if !ok {
				continue
			}

			endpointName := getString(opMap, "summary")
			if endpointName == "" {
				endpointName = getString(opMap, "operationId")
			}
			if endpointName == "" {
				endpointName = fmt.Sprintf("%s %s", methodUpper, path)
			}

			var params []models.Param
			var headers []models.Header

			if rawParams, ok := getSlice(opMap, "parameters"); ok {
				for _, raw := range rawParams {
					pMap, ok := raw.(map[string]interface{})
					if !ok {
						continue
					}
					pName := getString(pMap, "name")
					pIn := getString(pMap, "in")
					if pName == "" {
						continue
					}
					nv := models.NameValue{
						Name:    pName,
						Value:   "",
						Enabled: true,
						Type:    "plain",
					}
					switch pIn {
					case "query":
						params = append(params, nv)
					case "header":
						headers = append(headers, nv)
					}
				}
			}

			endpoints = append(endpoints, models.Endpoint{
				ID:        "e-" + uuid.NewString(),
				ServiceID: serviceID,
				Name:      endpointName,
				Method:    methodUpper,
				URL:       path,
				Metadata: models.EndpointMetadata{
					Version:     "1.0",
					LastUpdated: now,
				},
				Params:      params,
				Headers:     headers,
				Preflight:   nil,
				LastVersion: 0,
				Versions:    nil,
			})
		}
	}

	if len(endpoints) == 0 {
		return "", nil, fmt.Errorf("no endpoints found in spec")
	}

	return baseURL, endpoints, nil
}

// extractBaseURL extracts a base URL from an OpenAPI/Swagger spec map.
func extractBaseURL(spec map[string]interface{}) string {
	// Try OpenAPI 3 servers
	if servers, ok := getSlice(spec, "servers"); ok && len(servers) > 0 {
		if server, ok := servers[0].(map[string]interface{}); ok {
			if u := getString(server, "url"); u != "" {
				return u
			}
		}
	}

	// Try Swagger 2
	host := getString(spec, "host")
	scheme := "https"
	if schemes, ok := getSlice(spec, "schemes"); ok && len(schemes) > 0 {
		if s, ok := schemes[0].(string); ok {
			scheme = s
		}
	}
	basePath := getString(spec, "basePath")

	if host != "" {
		return fmt.Sprintf("%s://%s%s", scheme, host, basePath)
	}

	return "https://api.example.com"
}

// ----- Generic map helpers -----

func getMap(m map[string]interface{}, key string) (map[string]interface{}, bool) {
	v, ok := m[key]
	if !ok {
		return nil, false
	}
	result, ok := v.(map[string]interface{})
	return result, ok
}

func getSlice(m map[string]interface{}, key string) ([]interface{}, bool) {
	v, ok := m[key]
	if !ok {
		return nil, false
	}
	result, ok := v.([]interface{})
	return result, ok
}

func getString(m map[string]interface{}, key string) string {
	v, ok := m[key]
	if !ok {
		return ""
	}
	s, _ := v.(string)
	return s
}
